### PR TITLE
[FLINK-2243] [storm-compat] Added finite spout functionality to Storm compatibility layer

### PR DESCRIPTION
**Pull request by szape**
Added finite spout functionality to Storm compatibility layer

Storm spouts emit infinite streams. It is important to support the users of Flink's Storm compatibility layer by providing an interface that can add a finite nature to any storm spouts, given that Flink actually supports finite source streams.

### CODE REVIEW

File: flink-contrib/flink-storm-compatibility/flink-storm-compatibility-examples/src/main/java/org/apache/flink/stormcompatibility/util/FiniteStormFileSpout.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	}
+
+	/**
+	 * Can be called before nextTuple() any times including 0.
+	 */
+	public boolean reachedEnd() {
+		try {
+			readLine();
+		} catch (IOException e) {
+			e.printStackTrace();
```
Comment (rmetzger) [inline flink-contrib/flink-storm-compatibility/flink-storm-compatibility-examples/src/main/java/org/apache/flink/stormcompatibility/util/FiniteStormFileSpout.java]: Is this exception used to signal the end of the stream? or is it a error condition?
In any case, doing a `e.printStackStrace()` is not good because it will log it to the stdout file.
I would rethrow a runtime exception

File: flink-contrib/flink-storm-compatibility/flink-storm-compatibility-examples/src/main/java/org/apache/flink/stormcompatibility/wordcount/SpoutSourceWordCount.java [hunk 2]
```diff
@@ -145,7 +145,7 @@ private static boolean parseParameters(final String[] args) {
 		}
 
 		return env.addSource(new StormFiniteSpoutWrapper<String>(new StormInMemorySpout(WordCountData.WORDS), true),
-				TypeExtractor.getForClass(String.class));
+				TypeExtractor.getForClass(String.class)).setParallelism(1);
```
Comment (rmetzger) [inline flink-contrib/flink-storm-compatibility/flink-storm-compatibility-examples/src/main/java/org/apache/flink/stormcompatibility/wordcount/SpoutSourceWordCount.java]: Why do you fix the parallelism here?

File: flink-contrib/flink-storm-compatibility/flink-storm-compatibility-examples/src/main/java/org/apache/flink/stormcompatibility/util/FiniteStormFileSpout.java (hunk 1 above)
Comment (szape) [inline flink-contrib/flink-storm-compatibility/flink-storm-compatibility-examples/src/main/java/org/apache/flink/stormcompatibility/util/FiniteStormFileSpout.java]: Thanks for pointing it out! I fixed it.

File: flink-contrib/flink-storm-compatibility/flink-storm-compatibility-examples/src/main/java/org/apache/flink/stormcompatibility/wordcount/SpoutSourceWordCount.java (hunk 2 above)
Comment (szape) [inline flink-contrib/flink-storm-compatibility/flink-storm-compatibility-examples/src/main/java/org/apache/flink/stormcompatibility/wordcount/SpoutSourceWordCount.java]: I thought it was a mistake, because without setting the parallelism to 1, the input will be duplicated. It does not affect the ITCases directly, though, because they run in 'file input/output' mode, but it still seems out of order.
Any reason against setting the parallelism?

Comment (mjsax) [inline flink-contrib/flink-storm-compatibility/flink-storm-compatibility-examples/src/main/java/org/apache/flink/stormcompatibility/wordcount/SpoutSourceWordCount.java]: It is correct that the dop must be set to one. The used Spout is not parallelizable.

File: flink-contrib/flink-storm-compatibility/flink-storm-compatibility-examples/src/main/java/org/apache/flink/stormcompatibility/singlejoin/SingleJoinTopology.java [hunk 3]
```diff
@@ -1,90 +0,0 @@
-/*
```
Comment (mjsax) [inline flink-contrib/flink-storm-compatibility/flink-storm-compatibility-examples/src/main/java/org/apache/flink/stormcompatibility/singlejoin/SingleJoinTopology.java]: Why do you remove it? It's going to be fixed soon.

Comment (szape) [inline flink-contrib/flink-storm-compatibility/flink-storm-compatibility-examples/src/main/java/org/apache/flink/stormcompatibility/singlejoin/SingleJoinTopology.java]: Okay. I just didn't want a half-ready example to linger around. I will
remove that commit.

2015-06-22 11:58 GMT+02:00 Matthias J. Sax <notifications@github.com>:

Comment (mjsax) [inline flink-contrib/flink-storm-compatibility/flink-storm-compatibility-examples/src/main/java/org/apache/flink/stormcompatibility/singlejoin/SingleJoinTopology.java]: If you prepare an PR, limit the changes to the feature you are implementing. This example is not related to the new functionality, thus, you should just leave it as it is ;)

**Comment by mjsax** [comment]
Any news on this PR? Do you still work on it?

**Comment by szape** [comment]
No, it is finished and waiting for acceptance. As I remember, Márton wanted
to review the code, before merging it into the master.

Peter

2015-07-11 16:27 GMT+02:00 Matthias J. Sax <notifications@github.com>:

**Comment by mjsax** [comment]
Can you also extend documentation for this feature? (I followed a "only-1-sentence-in-1-line" approach; please do the same). I would suggest to open a new section "Flink extensions" before "Storm Compatibility Examples".

**Comment by szape** [comment]
Of couse, I will do it.

2015-07-22 14:31 GMT+02:00 Matthias J. Sax <notifications@github.com>:

**Comment by mbalassi** [comment]
If no objections I will merge this in the evening. (Maybe also squashing the commits a bit.)

**Comment by mjsax** [comment]
Nice :)
